### DISCUSSION TITLE
Welcome to nirs-toolbox Discussions!

### DISCUSSION THREAD

**Discussion by huppertt** [huppertt/nirs-toolbox#2] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by LeiaHealy** [comment]
Hi Dr. Huppert!

I just have a quick question regarding the differences between the probe 1020 and probe class. When performing image reconstruction, everything works fine until I run the ImageReconMFX module in which the below error message pops up. I think that the issue may be that my SubjStats contains a probe of the 1020 class, so its probe.link variable is different from the standard probe.link variable. Is there anyway to convert from one to the other? Or am I doing something wrong in the subject analysis portion to have my probe.link end up being the inherited probe1020 class rather then just a probe?

################
Error Message:

Error using tabular/ismember (line 37)
A and B must contain the same variables.

Error in nirs.modules.ImageReconMFX/runThis (line 114)
                [ia,ib]=ismember(thisprobe.link,S(idx).probe.link,'rows');        <----- comparing probe1020.link to probe.link

Error in nirs.modules.AbstractModule/run (line 77)
               out = obj.runThis( inputs );
################

Best,
Liam Healy

**Reply by xuetongzhai** [reply]
Hi Liam. I am a postdoc working in Dr. Huppert's lab. Can you send me your data and code to my email xuz19@pitt.edu. Then I can take a look and let you know. 

-Xuetong

**Comment by huppertt** [comment]
My best guess is that the stimulus you are creating is a struct and not the actual StimulusEvents class.

>> raw=nirs.testing.simARNoise;
>> stim=nirs.design.StimulusEvents;    % create empty stim class.   <<<<  I THINK YOU ARE MISSING THIS LINE
>> stim.onset=[30:30:300];   % events at 30,60,90…300s
>> stim.dur=10*ones(1,length(stim.onset));  % all events are 10s duration
>> stim.amp=ones(1,length(stim.onset));
>> raw.stimulus('MyTaskName')=stim;    % assign to the dictionary as “MyTaskName” as the condition
>> raw.draw;  % this should now show the task marks
>> stim.draw; % this should draw the event marks


My postdoc Hendrik (cc’d) can help if this doesn’t solve the issue for you.

-Ted



Theodore Huppert, PhD
University of Pittsburgh
Department of Electrical and Computer Engineering

Email: ***@***.******@***.***>
Phone: 1-412-383-5167
Website: www.huppertlab.net


From: FatimaSiba ***@***.***>
Reply-To: huppertt/nirs-toolbox ***@***.***>
Date: Tuesday, August 24, 2021 at 11:44 AM
To: huppertt/nirs-toolbox ***@***.***>
Cc: huppertt ***@***.***>, Author ***@***.***>
Subject: Re: [huppertt/nirs-toolbox] Welcome to nirs-toolbox Discussions! (#2)


Dear Dr. Huppertt and lab team members

I'm trying to process some practice for my experiment. I'm still in the beginning of the process. I'm editing the stimulus by creating my own dictionary class variable and assigning it to the raw data. When I visualize the data the stimulus looks correct. However when I try running the GLM model, I'm running into an error:

Warning: No stim conditions left. Did you provide the correct stim names?

Error using fields
Invalid input argument of type 'double'. Input must be a structure or a Java or COM object.

Error in nirs.modules.GLM/runThis (line 154)
if(~ismember('regressor_no_interest',fields(st)))

Error in nirs.modules.AbstractModule/run (line 77)
out = obj.runThis( inputs );

I could see in the GLM function the st variable is empty and that's why the error is occurring.
However, I'm not sure why that is. The stimulus names are correct in the raw.stimulus.key as well as raw.stimulus.values.names.

I appreciate your help

Thanks
Fatima Sibaii

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/huppertt/nirs-toolbox/discussions/2#discussioncomment-1229330>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALNU3OXOXALSXCXH5A4WRODT6PZCXANCNFSM4VFTEJPQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Reply by FatimaSiba** [reply]
Hey Dr. Huppertt 
Thank you, we figured out the issue and it is now resolved.  
Fatima Sibaii

**Comment by bbreeze25** [comment]
dear all, what version of Matlab would be the most stable to work with analyzIR? looks like I am getting errors that are Matlab version specific

**Comment by FatimaSiba** [comment]
Hello 

I have found a post from 2018 that discussed getting the residuals after running the GLM using advanded.nirs.modules.GLMResiduals. 
Here is the link: 
https://huppertlab.net/discussion/forum/topic/can-residuals-be-extracted-after-glm/
I have tried using this function, however I'm getting this error: 
Unable to resolve the name nirs.modules.GLMResiduals. 
I'm not sure what that error means. 
Also, if this were to work, are the residuals from the GLM considered "filtered time series"? 

Thanks 
Fatima

**Comment by huppertt** [comment]
I think the code was moved to advanced.nirs.modules.GLMresiduals.  The advanced namespace contains a bunch of stuff that is either work-in-progress or was part of a paper comparing methods and determined to be not the best method (e.g. some of the variations on how to use short-separation channels are in there after we found putting them in the GLM was the best approach).

The time-course data that will come back after this module will have the average response removed but hasn’t been really filtered beyond that unless you additionally have correction regressors in the GLM (e.g. short-separations or DCT/polynomial detrending terms).   E.g.  residual = Y – X*beta, so if X has “filtering” regressors, these will also be projected out of the residual.  Beta is estimated by the same AR-IRLS GLM model, but it goes back to the “raw” Y to compute the residuals once it finds beta

Theodore Huppert, PhD
Associate Professor
University of Pittsburgh
Department of Electrical and Computer Engineering

Email: ***@***.******@***.***>
Phone: 1-412-383-5167
Website: www.huppertlab.net<http://www.huppertlab.net>

Schedule a meeting with me at https://calendly.com/huppert1



From: Fatima Sibaii ***@***.***>
Date: Tuesday, February 22, 2022 at 12:18 AM
To: huppertt/nirs-toolbox ***@***.***>
Cc: huppertt ***@***.***>, Author ***@***.***>
Subject: Re: [huppertt/nirs-toolbox] Welcome to nirs-toolbox Discussions! (Discussion #2)

Hello

I have found a post from 2018 that discussed getting the residuals after running the GLM using advanded.nirs.modules.GLMResiduals.
Here is the link:
https://huppertlab.net/discussion/forum/topic/can-residuals-be-extracted-after-glm/
I have tried using this function, however I'm getting this error:
Unable to resolve the name nirs.modules.GLMResiduals.
I'm not sure what that error means.
Also, if this were to work, are the residuals from the GLM considered "filtered time series"?

Thanks
Fatima

—
Reply to this email directly, view it on GitHub<https://github.com/huppertt/nirs-toolbox/discussions/2#discussioncomment-2225889>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALNU3OWFUOIHOLBLZNWJBITU4MMB5ANCNFSM4VFTEJPQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by FatimaSiba** [reply]
Dear Dr. Huppert 

Thank you so much for your prompt reply! I changed the handle in MATLAB to match the new location of the GLMresiduals function, and it is now running. However, I'm running into an error at this line of code:  data(i).data = data(i).data - X * reshape(S(i).beta,[nchan ncond])'; 
I believe it's because the size of X depends on my conditions so it is of size: datapoints*conditions, while beta is reshaped to (numbers of channels*(betalength/number of channels)). I have 4 conditions (so X is timepointsx4), but beta has a length of 192, with 64 channels so it is reshaped to 3x64. Therefor, matrix multiplication is not working.
Also, the X created is all 0s, no values. 
I appreciate any tips on how to solve this error! 

Thank you 
Fatima

**Comment by DemianVera** [comment]
Hi, Dr. Huppert. I was watching [this webinar](https://www.youtube.com/watch?v=fBgbHgFubyI) and trying to reproduce the ROC curve analysis. I couldn't find the exact code you used in that webinar, so I managed to do it myself in Matlab (2021a), following the video and using the same data files. For any reason, when I compute the ROC curves, two problems appear: the first one is that there seems to be an error in ChannelStatsRoc.m, in Line 153; I suppose that where it says if(iscellstr(types(i))), it should say if(iscellstr(types(j))). Doing this replacement, it works (it doesn't complain, at least :-/ ). 
The second problem I find is that all ROC curves (for any pipeline, and for hbo, hbr and joint) are almost straight lines with an area under the curve of ~0.5. 
I'm using a recent version of nirs toolbox (I downloaded it last week from github). 

Thanks for your time and help, and sorry for bothering you with this.
I really appreciate all you're doing for the fNIRS community. It's a great, great job.

Best regards from Argentina,

Demián Vera.

**Reply by xuetongzhai** [reply]
Hi Demián,

Could you please share the code you are running? That would be helpful to debug. You can either drop the m file in the comment or send me via email at xuz19@pitt.edu. 

Thank you,
Xuetong

**Reply by DemianVera** [reply]
Hi, Xuetong. I sent my answer to you via email at xuz19@pitt.edu one week ago. Did you received it? Please let me know if you need more information or more details. Thanks!

Demián.

**Reply by xuetongzhai** [reply]
Hi Demián,

Sorry. I did not receive an email from you. I also checked the spam folder and nothing was found. Not sure what the problem is. Can you resend it or just attach the files in this comment?

Thank you,
Xuetong

**Reply by DemianVera** [reply]
Hi, Xuetong. Thanks for your quick answer; the code I'm running is, more or less, the one that Dr. Huppert shows in his webinar. See the .m attached in this email. I also include what I get (the two attached figures).
Those figure show the comparison between job1, job2 and job3. As you can see, I made some modifications to the ChannelStatsROC class, so the plots now have a title (in order to distinguish between ROC and type 1 error curves). The other modification I have made is, as I said in my first question: the line 153 of the original version of ChannelStatsRoc.m, because there was an index that was causing some problems (I replaced i by j).

I tried with other jobs and usually I get those straight lines in the ROC plots. It's like the methods are pretty bad, but that's not what the tutorial shows, neither what I expect. 

Thanks again for your time and kind answer.

Best regards,
Demián Vera.
[DrHuppertTutorial.txt](https://github.com/huppertt/nirs-toolbox/files/8334409/DrHuppertTutorial.txt)
 ![ChannelsStatsROC1](https://user-images.githubusercontent.com/60976791/159738368-c91224e5-9e28-48f2-a484-f34387873066.png)

**Reply by xuetongzhai** [reply]
Thank you, Demián. I will take a look and get back to you asap.

**Reply by DemianVera** [reply]
Hi, Xuetong. Sorry for bothering you again, did you have a second (or several seconds) to take a look to ChannelStatsROC.m? My colleagues and I tried to find what's wrong (if there's something wrong with the code), but couldn't find anything; the results we get keep differing from the results that Dr. Huppert gets in the tutorial.
Thanks for your time.
Best regards,

Demián.

**Reply by xuetongzhai** [reply]
Hi Demián,

Let me look at it and try to get back to you by this weekend.

**Reply by DemianVera** [reply]
Hi! Sorry for bothering you again; As I see no answer, I think there was no problem with the code and maybe I was wrong. Or maybe you had no time to check it out. Thanks for your time, anyway!

**Reply by nicdc** [reply]
Just to add, I've had the same issue using this script. It was working fine when I tested it previously (soon after Prof. Huppert gave the webinar at NIRx) but some subsequent change seems to have broken this script.

**Reply by nicdc** [reply]
Looked into this a bit more, I have a version of the toolbox from 17 months ago, when my TortoiseHg stopped updating the local repo due to some authentication issue. Reruning the same code that was shared in the demo on the same data with two versions of the local repo gives me two very different results. 
[simulations_17monthsago.pdf](https://github.com/huppertt/nirs-toolbox/files/10153622/simulations_17monthsago.pdf)
[simulations_today.pdf](https://github.com/huppertt/nirs-toolbox/files/10153623/simulations_today.pdf)

**Reply by DemianVera** [reply]
Thanks! Yes, they are very different. Something happened and no one noticed...

**Comment by XueruiPeng** [comment]
Dear Dr. Huppert,

Thanks a lot for your wonderful and useful toolbox.
I have some questions about parametric modulation analysis.

When I use nirs.design.split_parametric to manipulate my dataset, I have nan values in cond: amp.
 Specifically, here is my code to do so:
```
 stimTable = nirs.createStimulusTable(myRawdata);
 %% then I put amp value of my cond1 into the stimTable with some code %%
 job = nirs.modules.ChangeStimulusInfo();
 job.ChangeTable = stimTable;
 rawChanged = job.run(rawChanged);
rawChanged = nirs.design.split_parametric(rawChanged,'con1*(1+Amp)',true);
```
After running the code, I got a new stim named con1:amp, with the amp values being all nan. And when stats analysis is performed, this condition is automatically removed. Did I use the function incorrectly?

Then to try to solve this, I manually added the parametric modulation term into my data set. So I created a new cond named ‘_con1:amp_’ with the same onset and duration with _con1_ and mean-centered _amp_ value. Then perform the stats. Is it correct to do so?

Furthermore, I have two environment-related values in each trial (the two values are highly correlated, r = 0.73). So when I want to see the effect of one of them, I need to include another one in the model to regress its effect out. Can I add the second parametric modulation as I did for the first one? Then I have _con1_ (with amp all equal to 1), _con1:amp1_, _con1:amp2_. Should I have other steps to deal with their high correlation (such as orthogonalizing them beforehand)? Or, from my understanding, the regression model can help me deal with this kind of colinearity (I use the AR_IRLS model). Does the order of the two parametric modulated terms matter?

I appreciate any answers, tips, and suggestions about that.
Xuerui

**Comment by armandosajunior** [comment]
Hi guys,

Sorry if this is a silly question. I love the nirs toolbox, I'd love to continue using it, but I don't know how to remove bad channels in the toolbox - which is kind crucial to me. I usually do some preprocessing, glm and then linear mixed model. But if I try to remove bad channels, I usually get some errors, that I think is related to an uneven number of channels (e.g. Unable to perform assignment because the indices on the left side are not compatible with the size
of the right side.

Error in nirs.modules.MixedEffects/runThis (line 225)
            X(idx, :)   = X;

Error in nirs.modules.AbstractModule/run (line 73)
               out = obj.runThis( inputs );)".

Can anybody help me?

**Comment by DemianVera** [comment]
Hi everyone!
I use nirs toolbox to process my fNIRS data. Now I'm trying to open some data (BIDS) that contains data in snirf format. So I must use the nirs.io.loadSNIRF function; however, a function called snirf2data throws errors when trying to read the field SourcePos (line 60 of the present function). I think this happens because my data has the fields sourcePos3D and detectorPos3D. As I can see, this function takes that into account, but only partially. 
Can you help me, please? Is there any other way to load snirf data? 

Thanks in advance!

**Comment by AlinaSchulte** [comment]
Hi!

To identify channels with bad data quality, I previously used the scalp coupling index. In addition, I found the structure noise index (nirs.math.structnoiseindex) that is used within the QA report to identify bad channels. It seems to detect channels with large amplitudes (median deviations) that remain after removing autocorrelations. I'd like to understand a bit better what it is doing exactly. Has anyone else used it and could elaborate on its functioning?

Best regards,
Alina

**Comment by AlinaSchulte** [comment]
Hi,

I am wondering whether it is okay to use a band-pass filter before applying the AR-IRLS algorithm to my data. Within the code I see that low-pass filtering is not recommended (see screenshot below) and can lead to unstable AR coefficients. I could not figure out what that means for the data and the glm results, but trying my filter on a simulated dataset does not seem to cause visible problems and helps finding the simulated response.

Looking into the literature, I see that in some publications filtering is used with the AR-IRLS glm model estimation. That's also the case for the paper by Santosa et al., 2020. Here however, the reversed problem is mentioned (high-pass filtering being problematic).

![image](https://user-images.githubusercontent.com/80465276/207373550-e0a9ff35-2eb1-4f96-89df-58b2542f774b.png)

Can anyone help with this?

Best,
Alina

**Comment by AMGoodwill** [comment]
Hi, I am wondering if the connectivity functions in nirs toolbox could be used for task-based as well to examine dynamic networks?
Thank you! 
Alicia

**Comment by Raya2023** [comment]
Hi,

Thank you for sharing your toolbox! @huppertt @xuetongzhai @hendriksantosa 
How can we get support for our inquiries or the issues occurs when using NIRS toolbox?
I'm facing a problem when loading my .nirs files, some of them are loaded with missing stimulus table or missing column, although I confirmed that their original csv files (acquired by Hitachi) include the mark values. And some other files are loaded fine (converted using the same script and loaded using the same function as the files with missing stimulus). What would be the solution to get my stimuli values right?
BTW, when I tried loadHitachi function for directly loading Hitachi csv files, MATLAB hangs.

Thank you in advance.

**Comment by shatton412** [comment]
Hello, 
I had a concept question to check my knowledge before I start my analysis. I am interested in doing multimodal FNIRS and EEG data fusion. I know you can convolve the eeg signal with the HRF from the fnirs using eeg.modules.HRFConvolve. However, my data set is a resting data set. Most of the papers I read make it seem that most HRF's can only be extracted from task-based experiments. Am I able to use eeg.modules.HRFConvolve on my resting data? Or is this only feasible on task-related paradigms?  
Thank you in advance!

**Comment by pimbrouwer** [comment]
Hello Dr. Huppert!

Thank you for the NIRS toolbox. I wanted to explore the possibilities and was trying to run the fnirs_example_shortsep.m demo. The dataset which is used in this demo can't be found at huppertlab.net anymore. Could you share the NIRx_data_SS.zip with me?

Thank you!

**Comment by ChristophGeissler** [comment]
Dear Dr. Huppert,

First, I want to thank you and your team! I’ve been using the Nirs-Toolbox almost since it came out in 2018. To my shame, I must confess, I hadn’t really updated for a long time. However, when switching to a Windows 11 PC this fall, I took this as an opportunity to adapt the newest version of the toolbox. I recently reran an analysis I had originally made with my old PC (and the old version of the toolbox). For some reason, the first level analysis (AR-IRLS) now uses about twice as much memory as before, but more gravely, the results completely changed. Do you have any idea what might be a cause of this?
Thank you and best regards,

Christoph
